File: src/forecast_warden/synthetic.py
"""Generate ~14 days of synthetic zone metrics (some intentionally sick)."""

from __future__ import annotations

import math
import random
from datetime import datetime, timedelta, timezone

from forecast_warden.models import BaselineStats, MetricRow

ZONES = ["Z1", "Z2", "Z3", "Z4", "Z5", "Z6", "Z7", "Z8"]

# Intentionally sick zones on the latest run (for demo / golden path).
SICK_PROFILES: dict[str, dict] = {
    "Z3": {"mape": 0.42, "bias": 0.22, "coverage_80": 0.55, "n_actuals": 80},  # critical MAPE + bias + cov
    "Z7": {"mape": 0.31, "bias": -0.18, "coverage_80": 0.72, "n_actuals": 60},  # warning MAPE + bias
    "Z5": {"mape": 0.12, "bias": 0.02, "coverage_80": 0.81, "n_actuals": 12},  # low support only
}


def generate_metrics(
    days: int = 14,
    end_date: datetime | None = None,
    seed: int = 42,
) -> list[MetricRow]:
    rng = random.Random(seed)
    if end_date is None:
        end_date = datetime(2026, 9, 12, 6, 0, 0, tzinfo=timezone.utc)
    rows: list[MetricRow] = []
    for d in range(days):
        day = end_date - timedelta(days=days - 1 - d)
        run_id = day.strftime("%Y-%m-%d")
        generated_at = day
        for zone in ZONES:
            if d == days - 1 and zone in SICK_PROFILES:
                p = SICK_PROFILES[zone]
                rows.append(
                    MetricRow(
                        run_id=run_id,
                        entity_type="zone",
                        entity_id=zone,
                        mape=p["mape"],
                        bias=p["bias"],
                        coverage_80=p["coverage_80"],
                        n_actuals=p["n_actuals"],
                        generated_at=generated_at,
                    )
                )
                continue
            # Healthy-ish noise; last day kept especially calm so only SICK_PROFILES fire
            mape = max(0.05, min(0.20, 0.12 + rng.gauss(0, 0.02)))
            bias_sigma = 0.015 if d == days - 1 else 0.03
            bias_cap = 0.05 if d == days - 1 else 0.10
            bias = max(-bias_cap, min(bias_cap, rng.gauss(0, bias_sigma)))
            coverage = max(0.70, min(0.95, 0.82 + rng.gauss(0, 0.03)))
            n = int(max(40, min(120, 70 + rng.gauss(0, 15))))
            rows.append(
                MetricRow(
                    run_id=run_id,
                    entity_type="zone",
                    entity_id=zone,
                    mape=round(mape, 6),
                    bias=round(bias, 6),
                    coverage_80=round(coverage, 6),
                    n_actuals=n,
                    generated_at=generated_at,
                )
            )
    return rows


def compute_baseline(metrics: list[MetricRow], window_days: int = 28) -> list[BaselineStats]:
    """Rolling mean/std from all but the last run_id (so latest can z-score)."""
    if not metrics:
        return []
    run_ids = sorted({m.run_id for m in metrics})
    history_ids = set(run_ids[:-1]) if len(run_ids) > 1 else set(run_ids)
    by_entity: dict[str, list[MetricRow]] = {}
    for m in metrics:
        if m.run_id not in history_ids:
            continue
        by_entity.setdefault(m.entity_id, []).append(m)

    out: list[BaselineStats] = []
    for entity_id, rows in sorted(by_entity.items()):
        mapes = [r.mape for r in rows]
        biases = [r.bias for r in rows]
        out.append(
            BaselineStats(
                entity_type="zone",
                entity_id=entity_id,
                mape_mean=_mean(mapes),
                mape_std=_std(mapes),
                bias_mean=_mean(biases),
                bias_std=_std(biases),
                window_days=window_days,
            )
        )
    return out


def _mean(xs: list[float]) -> float:
    return sum(xs) / len(xs) if xs else 0.0


def _std(xs: list[float]) -> float:
    if len(xs) < 2:
        return 0.01  # avoid zero-division in z-score
    m = _mean(xs)
    var = sum((x - m) ** 2 for x in xs) / (len(xs) - 1)
    return max(math.sqrt(var), 0.01)

File: internal/synthetic/synthetic.go
package synthetic

import (
	"math"
	"math/rand"
	"sort"
	"time"

	"github.com/luizpedrini/forecast-warden/internal/models"
)

var Zones = []string{"Z1", "Z2", "Z3", "Z4", "Z5", "Z6", "Z7", "Z8"}

// SickProfiles intentionally sick zones on the latest run (demo / golden path).
var SickProfiles = map[string]struct {
	MAPE       float64
	Bias       float64
	Coverage80 float64
	NActuals   int
}{
	"Z3": {MAPE: 0.42, Bias: 0.22, Coverage80: 0.55, NActuals: 80},
	"Z7": {MAPE: 0.31, Bias: -0.18, Coverage80: 0.72, NActuals: 60},
	"Z5": {MAPE: 0.12, Bias: 0.02, Coverage80: 0.81, NActuals: 12},
}

func GenerateMetrics(days int, endDate time.Time, seed int64) []models.MetricRow {
	if days <= 0 {
		days = 14
	}
	if endDate.IsZero() {
		endDate = time.Date(2026, 9, 12, 6, 0, 0, 0, time.UTC)
	}
	rng := rand.New(rand.NewSource(seed))
	var rows []models.MetricRow
	for d := 0; d < days; d++ {
		day := endDate.AddDate(0, 0, -(days - 1 - d))
		runID := day.Format("2006-01-02")
		generatedAt := day
		for _, zone := range Zones {
			if d == days-1 {
				if p, ok := SickProfiles[zone]; ok {
					rows = append(rows, models.MetricRow{
						RunID:       runID,
						EntityType:  "zone",
						EntityID:    zone,
						MAPE:        p.MAPE,
						Bias:        p.Bias,
						Coverage80:  p.Coverage80,
						NActuals:    p.NActuals,
						GeneratedAt: generatedAt,
					})
					continue
				}
			}
			mape := clamp(0.12+rng.NormFloat64()*0.02, 0.05, 0.20)
			biasSigma := 0.03
			biasCap := 0.10
			if d == days-1 {
				biasSigma = 0.015
				biasCap = 0.05
			}
			bias := clamp(rng.NormFloat64()*biasSigma, -biasCap, biasCap)
			coverage := clamp(0.82+rng.NormFloat64()*0.03, 0.70, 0.95)
			n := int(clamp(70+rng.NormFloat64()*15, 40, 120))
			rows = append(rows, models.MetricRow{
				RunID:       runID,
				EntityType:  "zone",
				EntityID:    zone,
				MAPE:        round6(mape),
				Bias:        round6(bias),
				Coverage80:  round6(coverage),
				NActuals:    n,
				GeneratedAt: generatedAt,
			})
		}
	}
	return rows
}

func ComputeBaseline(metrics []models.MetricRow, windowDays int) []models.BaselineStats {
	if windowDays <= 0 {
		windowDays = 28
	}
	if len(metrics) == 0 {
		return nil
	}
	runSet := map[string]struct{}{}
	for _, m := range metrics {
		runSet[m.RunID] = struct{}{}
	}
	runIDs := make([]string, 0, len(runSet))
	for id := range runSet {
		runIDs = append(runIDs, id)
	}
	sort.Strings(runIDs)
	history := map[string]struct{}{}
	if len(runIDs) > 1 {
		for _, id := range runIDs[:len(runIDs)-1] {
			history[id] = struct{}{}
		}
	} else {
		for _, id := range runIDs {
			history[id] = struct{}{}
		}
	}
	byEntity := map[string][]models.MetricRow{}
	for _, m := range metrics {
		if _, ok := history[m.RunID]; !ok {
			continue
		}
		byEntity[m.EntityID] = append(byEntity[m.EntityID], m)
	}
	entityIDs := make([]string, 0, len(byEntity))
	for id := range byEntity {
		entityIDs = append(entityIDs, id)
	}
	sort.Strings(entityIDs)
	out := make([]models.BaselineStats, 0, len(entityIDs))
	for _, entityID := range entityIDs {
		rows := byEntity[entityID]
		mapes := make([]float64, len(rows))
		biases := make([]float64, len(rows))
		for i, r := range rows {
			mapes[i] = r.MAPE
			biases[i] = r.Bias
		}
		out = append(out, models.BaselineStats{
			EntityType: "zone",
			EntityID:   entityID,
			MAPEMean:   mean(mapes),
			MAPEStd:    std(mapes),
			BiasMean:   mean(biases),
			BiasStd:    std(biases),
			WindowDays: windowDays,
		})
	}
	return out
}

func mean(xs []float64) float64 {
	if len(xs) == 0 {
		return 0
	}
	s := 0.0
	for _, x := range xs {
		s += x
	}
	return s / float64(len(xs))
}

func std(xs []float64) float64 {
	if len(xs) < 2 {
		return 0.01
	}
	m := mean(xs)
	variance := 0.0
	for _, x := range xs {
		d := x - m
		variance += d * d
	}
	variance /= float64(len(xs) - 1)
	return math.Max(math.Sqrt(variance), 0.01)
}

func clamp(x, lo, hi float64) float64 {
	if x < lo {
		return lo
	}
	if x > hi {
		return hi
	}
	return x
}

func round6(x float64) float64 {
	return math.Round(x*1e6) / 1e6
}

func DefaultEndDate() time.Time {
	return time.Date(2026, 9, 12, 6, 0, 0, 0, time.UTC)
}
